src/coprocessor/dag/aggr_fn/impl_max_min.rs

```diff
@@ -0,0 +1,342 @@
+
+        // `MAX/MIN` outputs one column which has the same type with its child
+        out_schema.push(aggr_def.take_field_type());
+
+        let child = aggr_def.take_children().into_iter().next().unwrap();
+        let eval_type = EvalType::try_from(child.get_field_type().tp()).unwrap();
+        out_exp.push(RpnExpressionBuilder::build_from_expr_tree(
+            child,
+            time_zone,
+            src_schema.len(),
+        )?);
+
+        match_template_evaluable! {
+            TT, match eval_type {
+                EvalType::TT => Ok(Box::new(AggFnExtremum::<TT>::new(T::ORD)))
+            }
+        }
+    }
+}
+
+/// The bit operation aggregate functions.
+#[derive(Debug, AggrFunction)]
+#[aggr_function(state = AggFnStateExtremum::<T>::new(self.ord))]
+pub struct AggFnExtremum<T>
+where
+    T: Evaluable + Ord,
+    VectorValue: VectorValueExt<T>,
+{
+    _phantom: std::marker::PhantomData<T>,
+    ord: Ordering,
+}
+
+impl<T> AggFnExtremum<T>
+where
+    T: Evaluable + Ord,
+    VectorValue: VectorValueExt<T>,
+{
+    fn new(ord: Ordering) -> Self {
+        Self {
+            _phantom: std::marker::PhantomData,
+            ord,
+        }
+    }
+}
+
+/// The state of the MAX/MIN aggregate function.
+#[derive(Debug)]
+pub struct AggFnStateExtremum<T>
+where
+    T: Evaluable + Ord,
+    VectorValue: VectorValueExt<T>,
+{
+    _phantom: std::marker::PhantomData<T>,
+    ord: Ordering,
+    extremum: Option<T>,
+}
+
+impl<T> AggFnStateExtremum<T>
+where
+    T: Evaluable + Ord,
+    VectorValue: VectorValueExt<T>,
+{
+    pub fn new(ord: Ordering) -> Self {
+        Self {
+            _phantom: std::marker::PhantomData,
+            ord,
+            extremum: None,
+        }
+    }
+}
+
+impl<T> super::ConcreteAggrFunctionState for AggFnStateExtremum<T>
+where
+    T: Evaluable + Ord,
+    VectorValue: VectorValueExt<T>,
+{
+    type ParameterType = T;
+
+    #[inline]
+    fn update_concrete(
+        &mut self,
+        _ctx: &mut EvalContext,
+        value: &Option<Self::ParameterType>,
+    ) -> Result<()> {
+        if value.is_some() && (self.extremum.is_none() || self.extremum.cmp(value) == self.ord) {
+            self.extremum = value.clone();
+        }
+        Ok(())
+    }
+
+    #[inline]
+    fn push_result(&self, _ctx: &mut EvalContext, target: &mut [VectorValue]) -> Result<()> {
+        target[0].push(self.extremum.clone());
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use cop_datatype::EvalType;
+    use tipb_helper::ExprDefBuilder;
+
+    use super::*;
+    use crate::coprocessor::codec::batch::{LazyBatchColumn, LazyBatchColumnVec};
+    use crate::coprocessor::dag::aggr_fn::parser::AggrDefinitionParser;
+    use crate::coprocessor::dag::aggr_fn::AggrFunction;
+    use cop_datatype::{FieldTypeAccessor, FieldTypeTp};
+
+    #[test]
+    fn test_max() {
+        let mut ctx = EvalContext::default();
+        let function = AggFnExtremum::<Int>::new(Max::ORD);
```

Comment:
The construction looks confusing. ORD should not be exposed.

Reply:
addressed